### PR TITLE
Add documentation for out-of-cluster sample-apiserver development

### PR DESCRIPTION
**Pull request by mvladev**
**What this PR does / why we need it**:
It shows how a developer can run and develop the `sample-apiserver` out-of-cluster. This improves development cycle without the need to build a binary, build a docker image, push it to registry and then run it on your local `minikube`.

**Which issue(s) this PR fixes** none

**Special notes for your reviewer**: I'm not sure, if it's better to split the guide in 2 parts with multiple `.md` files.

**Release note**: 

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md [hunk 1]
```diff
@@ -1,56 +1,67 @@
...
 ## Clone the repository
 
-In order to build the sample apiserver image we will need to build the apiserver binary.
+In order to build or develop the sample apiserver image we'll need to clone the repository.
 
-```
+```bash
 cd $GOPATH/src
 mkdir -p k8s.io
 cd k8s.io
```
Comment (php-coder) [inline staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: These 3 commands could be simplified to
```bash
mkdir $GOPATH/src/k8s.io
cd $GOPATH/src/k8s.io
```

Comment (mvladev) [reply staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: @php-coder nice catch. I'll fix it in subsequent PR

**Comment by sttts** [comment]
Related: https://github.com/kubernetes/kubernetes/pull/55476

Haven't read through this PR yet. Maybe my changes are not necessary.

**Comment by mvladev** [comment]
@sttts I was not aware of that PR. I would say that they both cover the same local development scenario, but have different use-cases: 

- #55476's focus is only on `sample-apiserver` development, without the aggregation layer. You should use it, If you don't need any of the `core` api groups (`core/v1`, `apps/v1`, etc...) in your admission plugins. It runs with local `etcd`, generates certificates and requires a manually created `kubeconfig` to use with `kubectl`.
- this PR's focus `sample-apiserver` development, with aggregation layer. You should use it, If you need any of the `core` api groups (`core/v1`, `apps/v1`, etc...) in your admission plugins or write custom controller that requires both groups. It runs with `minikube` hosted `etcd` and reuses minikube's certificates and `kubeconfig`

IMHO both uses-cases are valuable and coexist side-by-side.

File: staging/src/k8s.io/sample-apiserver/artifacts/example/local-development/etcd-service.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: wardle-etcd
```
Comment (sttts) [inline staging/src/k8s.io/sample-apiserver/artifacts/example/local-development/etcd-service.yaml]: maybe rename this to etcd? Like the file. The namespace is wardle already.

File: staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md [hunk 3]
```diff
@@ -1,56 +1,67 @@
 # Minikube walkthrough
 
 This document will take you through setting up and trying the sample apiserver on a local minikube from a fresh clone of this repo.
+You can optionally run and develop the sample apiserver on your local machine and use minikube to host the etcd.
```
Comment (sttts) [inline staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: and use minikube only to host the etcd.

Comment (sttts) [reply staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: But why don't you run etcd locally as well? Isn't that much easier?

File: staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md [hunk 4]
```diff
@@ -1,56 +1,67 @@
...
 
-## Pre requisites
+## Prerequisites
 
-- Go 1.7.x or later installed and setup. More infomration can be found at [go installation](https://golang.org/doc/install)
+- Go 1.7.x or later installed and setup. More information can be found at [go installation](https://golang.org/doc/install)
```
Comment (sttts) [inline staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: Honestly, I would drop the whole sentence. Having Go installed for a Go project is pretty obvious. Moreover, Go 1.7.x. is probably pretty outdated. Not even sure it still works with our libraries.

File: staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md [hunk 5]
```diff
@@ -1,56 +1,67 @@
...
-```
+From the root of this repo, where `main.go` is located, run the following command:
+
+```bash
 export GOOS=linux; go build .
 ```
-if everything went well, you should have a binary called ```sample-apiserver``` present in your current directory.
 
-## Build the container image
+if everything went well, you should have a binary called `sample-apiserver` present in your current directory.
```
Comment (sttts) [inline staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: If

File: staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md [hunk 6]
```diff
@@ -86,18 +97,114 @@ kubectl create -f artifacts/example/service.yaml -n wardle
 kubectl create -f artifacts/example/apiservice.yaml
 ```
 
-## Test that your setup has worked
+You can skip to [test your setup](#test-that-your-setup-has-worked)
```
Comment (sttts) [inline staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: You can continue with

File: staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md [hunk 7]
```diff
@@ -86,18 +97,114 @@ kubectl create -f artifacts/example/service.yaml -n wardle
...
+    metadata:
+      labels:
+        etcd: "true"
+    spec:
+      containers:
+      - image: quay.io/coreos/etcd:v3.1.10
+        name: etcd
+        command:
+        - etcd
+        - -advertise-client-urls=http://${MINIKUBE_IP}:2379
```
Comment (sttts) [inline staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: do we really need this copy of the file here? Can't we describe how to replace the IP in the provided file?

**Comment by sttts** [comment]
Some comments. Looks good overall.

File: staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md (hunk 3 above)
Comment (mvladev) [reply staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: It's another binary you need to install and put in your `PATH`. I find it easier to run in Docker or Kubernetes.

File: staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md (hunk 7 above)
Comment (mvladev) [reply staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: You are referring to sample apisever + etcd replicationcontroller [manifest file](https://github.com/kubernetes/kubernetes/blob/7652c252d4ce2aea1563853bef0438769c5781fd/staging/src/k8s.io/sample-apiserver/artifacts/example/rc.yaml) or copy pasting into the terminal?

Comment (sttts) [reply staging/src/k8s.io/sample-apiserver/docs/minikube-walkthrough.md]: The former.

**Pull request closed without merging**